src/Passes/Source/Commandline/Tests/Unit/configuration.cpp

```diff
@@ -0,0 +1,151 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#include "AllocationManager/AllocationManager.hpp"
+#include "Commandline/ConfigurationManager.hpp"
+#include "gtest/gtest.h"
+
+using namespace microsoft::quantum;
+
+namespace
+{
+
+class TestConfig1
+{
+  public:
+    void setup(ConfigurationManager& config)
+    {
+        config.setSectionName("Base configuration", "");
+        config.addParameter(param1_, "param1", "");
+        config.addParameter(param2_, "param2", "");
+        config.addParameter(param3_, "param3", "");
+    }
+
+    bool param1() const
+    {
+        return param1_;
+    }
+
+    std::string param2() const
+    {
+        return param2_;
+    }
+
+    int32_t param3() const
+    {
+        return param3_;
+    }
+
+  private:
+    bool        param1_{false};
+    std::string param2_{""};
+    int32_t     param3_{9};
+};
+
+class TestConfig2
+{
+  public:
+    void setup(ConfigurationManager& config)
+    {
+        config.setSectionName("Base configuration", "");
+        config.addParameter(param1_, "param1", "");
+        config.addParameter(param2_, "param2", "");
+        config.addParameter(param3_, "param3", "");
+    }
+
+    bool param1() const
+    {
+        return param1_;
+    }
+
+    std::string param2() const
+    {
+        return param2_;
+    }
+
+    int32_t param3() const
+    {
+        return param3_;
+    }
+
+  private:
+    bool        param1_{true};
+    std::string param2_{"xxxx"};
+    int32_t     param3_{9};
+};
+
+} // namespace
+
+TEST(CommandlineTestSuite, Configuration)
+{
+    {
+        ConfigurationManager configuration_manager;
+        configuration_manager.addConfig<TestConfig1>();
+
+        ParameterParser parser;
+        configuration_manager.setupArguments(parser);
+        char* args[] = {"main", "--param1", "--param2", "hello", "--param3", "1337"};
+        parser.parseArgs(6, args);
+
+        configuration_manager.configure(parser);
```

Comment:
(Feel free to ignore)
Without diving into the code, it seems suspicious that for the same instance `configuration_manager` two different calls are done - `setupArguments()` and `configure()`, _with the same argument `parser`_ (rather than getting the same result with a single call).

Reply:
The first call configures the parser to tell it which arguments are boolean flags and the second call extracts the parsed information from the parser to set each of the configuration sections. Maybe a refactor of the function names would make it easier to understand this? I am happy to receive suggestions.

Reply:
Let's resolve this right now and treat it as something to keep in mind if there will be any refactoring of this mechanism.